proposals/pattern-const-expressions.md

```diff
@@ -0,0 +1,145 @@
+> - **`is` expressions with only the following subpatterns:**
+>   - **Boolean literal patterns.**
+>   - **Numeric literal patterns.**
+>   - **Character literal patterns.**
+>   - **String literal patterns.**
+>   - **Relative numeric literal patterns.**
+>   - **Relative character literal patterns.**
+>   - **Null literal patterns.**
+>   - **Default literal patterns.**
+>   - **And, or and not patterns.**
+>   - **Parenthesized patterns.**
+>   - **References to constant fields or locals inside constant patterns.**
+
+### Grammar
+[grammar]: #grammar
+The grammar remains untouched, as nothing changes syntactically for this feature.
+
+### Semantics
+[semantics]: #semantics
+A pattern expression only consisting of constant subpatterns with a constant LHS may be evaluated at compile time currently, but it is not considered as constant. With this change, the following adjustments enable constant pattern expressions to be used in constant contexts, like assigning a constant field or local, a default parameter, the condition in a constant conditional expression, etc.
+
+### Examples
+[examples]: #examples
+The below example shows pattern expressions being assigned to constant fields:
+```csharp
+public const int A = 4;
+public const bool B = A is 4;
+public const bool C = A is not 3 and <= 4 or 6;
+
+// DeploymentEnvironment is an enum type
+public const DeploymentEnvironment Environment = DeploymentEnvironment.Production;
+public const bool D = Environment
+    is DeploymentEnvironment.Production
+    or DeploymentEnvironment.Test;
+```
+
+In the expression assigned to `C`, since `A` is constant, and all the subpatterns on the pattern are constant, the entire expression only consists of constants. This result will be evaluated during compile-time and then be assigned to the constant field.
+
+Likewise, `Environment` is constant as an enum value, and so are the other subpatterns of the expression assigned to `D`.
+
+Another example, using `null` and default value checks:
+```csharp
+const int a = 4;
+const bool b = false;
+const long c = 4;
+const string d = "hello";
+
+const bool x = a is default(int); // always false
+const bool y = b is default(bool); // always true
+const bool z = d is not null; // always true
+const bool p = b is null; // always false
+```
+
+All the above are currently valid pattern matching expressions, that also emit warnings about their constant evaluation results, about them being always true or false.
+
+When assigning those expressions to a constant symbol, these warnings about the constant result of the expression will **not** be reported, as the user intends to capture the constant value of the expression.
+
+Pattern expressions containing non-constant subpatterns, like accessing properties, list patterns and var patterns, are **not** constant. In the below examples, all expressions will report compiler errors:
+
+```csharp
+const bool q = d is string { Length: 5 }; // Error: not a constant expression
```

Comment:
This is an interesting and special case for string. Can we treat length of string literal constant in some way? It can also be useful elsewhere.

Reply:
That falls as a separate concern and would suit for another proposal I believe.

Reply:
Yeah, _way_ out of scope of this proposal.